### ISSUE TITLE
Debugging and benchmarking libutils

### ISSUE THREAD

**Issue by litchipi** [OP-TEE/optee_os#4482]
Hello, after experimenting with OP-TEE on the stm32mp157c-ev1, it seams my xtest 1006 is very slow (~ 10 mins), and I'd like to analyse which part of the code is slow. I've traced the actual working code to the file `lib/libutils/isoc/bget.c`.

I have a working macro (on host with <time.h>) that outputs me a message along with a time difference since the last message, the problem is that I cannot find a way to import a function that returns me a time in OP-TEE. The unit of time doesn't matter as long as it is little enough (~ clock tick / nanoseconds).
I've found the `barrier_read_cntpct()` function, but I cannot find a way to get it compiled along with the libutils lib, as it relies on generated header files.

Am I missing an existing util ? Is there a way for me to use the function `barrier_read_cntpct()` inside libutils code ?
Thank you in advance

**TL;DR**: Need functions that returns time usable inside libutils for debug & benchmarking purpose

**Comment by jforissier** [comment]
Hi @litchipi,

The fact that `<arm_user_sysreg.h>` can't be included without errors in e.g., `bget.c` is quite unfortunate and I would describe this as a bug in our build environment. `libutils` is built three times with different compiler flags and different source files, for use by the TEE core, `ldelf` and TAs. It seems to me that the user sysregs should be accessible in all three contexts.

I think the issue is with the include flags in the various `sub.mk` files under `lib/libutils`. I'll take a closer look.

**Comment by litchipi** [comment]
It seams that the core needs a first pass of the compiler in order to generate the `out/arm/include/generated/arm32_user_sysreg.h` file included in `arm.h` I include in my code.
Is there some documentation about the whole compilation process ? I have to admit being a bit lost in it.

When trying to use the __KERNEL__ flag to define a blank function and an `extern` function pointing to a piece of code inside the core itself, it seams that the bget.c is compiled in a userland context (the flag is not set).

**Comment by jforissier** [comment]
OK things are a bit complicated and not totally consistent I think but I don't have any sane proposal to rework this at the moment.
However I think the following should help you address your immediate need which is calling `read_cntpct()` from `bget.c`. Please try the following patch. Builds for me.
```diff
diff --git a/lib/libutils/isoc/bget.c b/lib/libutils/isoc/bget.c
index 48ee1100..090f3c52 100644
--- a/lib/libutils/isoc/bget.c
+++ b/lib/libutils/isoc/bget.c
@@ -468,6 +468,11 @@
 extern char *sprintf();               /* Sun includes don't define sprintf */
 #endif
 
+#ifdef __KERNEL__
+#include <arm.h>
+#else
+#include <arm_user_sysreg.h>
+#endif
 #include <assert.h>
 #include <memory.h>
 
@@ -581,6 +586,7 @@ static bufsize buf_get_pos(struct bfhead *bf, bufsize align, bufsize hdr_size,
 {
     unsigned long buf = 0;
     bufsize pos = 0;
+    uint64_t t = read_cntpct();
 
     if (bf->bh.bsize < size)
        return -1;
diff --git a/lib/libutils/sub.mk b/lib/libutils/sub.mk
index e49bd899..442070b6 100644
--- a/lib/libutils/sub.mk
+++ b/lib/libutils/sub.mk
@@ -1,2 +1,4 @@
 subdirs-$(CFG_LIBUTILS_WITH_ISOC) += isoc
 subdirs-y += ext
+FORCE-GENSRCldelf: $(out-dir)/include/generated/arm32_user_sysreg.h
+cppflags-lib-y += -I$(out-dir)/include/generated
```

**Comment by litchipi** [comment]
I got the following error:
```
E/TC:? 0
E/TC:? 0 User mode undef-abort at address 0x400806b8
E/TC:? 0  fsr 0x00000000  ttbr0 0x2ffd3020  ttbr1 0x00000000  cidr 0x0
E/TC:? 0  cpu #1          cpsr 0x20000130
E/TC:? 0  r0 0x4008e059      r4 0x00000000    r8 0x400a50c4   r12 0x400a5140
E/TC:? 0  r1 0x4008e021      r5 0x4008e059    r9 0x40011750    sp 0x40011680
E/TC:? 0  r2 0x400115bb      r6 0x400a1048   r10 0x00000005    lr 0x4008f925
E/TC:? 0  r3 0x400115bc      r7 0x400116d0   r11 0x40185b68    pc 0x400806b8
E/LD:  Status of TA 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b
E/LD:   arch: arm
E/LD:  region  0: va 0x40002000 pa 0x00000000 size 0x002000 flags rw-s (ldelf)
E/LD:  region  1: va 0x40004000 pa 0x00000000 size 0x005000 flags r-xs (ldelf)
E/LD:  region  2: va 0x40009000 pa 0x00000000 size 0x001000 flags rw-s (ldelf)
E/LD:  region  3: va 0x4000a000 pa 0x00000000 size 0x004000 flags rw-s (ldelf)
E/LD:  region  4: va 0x4000e000 pa 0x00000000 size 0x001000 flags r--s
E/LD:  region  5: va 0x4000f000 pa 0x00000000 size 0x003000 flags rw-s (stack)
E/LD:  region  6: va 0x40012000 pa 0xf9218c04 size 0x001000 flags rw-- (param)
E/LD:  region  7: va 0x4007c000 pa 0x00000000 size 0x009000 flags r-xs [1]
E/LD:  region  8: va 0x40085000 pa 0x00008000 size 0x002000 flags rw-s [1]
E/LD:  region  9: va 0x4008e000 pa 0x00001000 size 0x017000 flags r-xs [0]
E/LD:  region 10: va 0x400a5000 pa 0x00018000 size 0x0e6000 flags rw-s [0]
E/LD:   [0] 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b @ 0x4008e000
E/LD:   [1] ffd2bded-ab7d-4988-95ee-e4962fff7154 @ 0x4007c000
regression_1000.c:566: TEEC_InvokeCommand(&session, 5, &op, &ret_orig) has an unexpected value: 0xffff3024 = TEE_ERROR_TARGET_DEAD, expected 0x0 = TEEC_SUCCESS
```
Is xtest (user-land secure world app) allowed to read from internal registers like that ? Maybe there's an authorization to get or an interdiction to remove somewhere in the regression testing suite... 
I've tried to hack it using a piece of code to read this register in `core/arch/arm/kernel/boot.c` (because why not), and declare it only in bget.h, but as bget is compiled without the kernel (__KERNEL__ is not defined), it's not building.

**Comment by jforissier** [comment]
> Is xtest (user-land secure world app) allowed to read from internal registers like that ? Maybe there's an authorization to get or an interdiction to remove somewhere in the regression testing suite...

Ah, right, you need to remove this line: https://github.com/OP-TEE/optee_os/blob/3.12.0/core/arch/arm/kernel/thread.c#L1187

**Comment by litchipi** [comment]
When I commented the condition, it didn't work better, however I succeeded to make it work by setting `CFG_FTRACE_SUPPORT=y`.
Because I don't really want  to bundle ftrace here, I tried to analyse what ftrace does that I don't, and in `optee_os/core/arch/arm/arm.mk` ftrace forces cflags for arm instructions instead of thumb (which could cause the problem as we inject asm code directly) but it still doesn't work, same error.

**Comment by litchipi** [comment]
Could the problem be in the file `lib/libutils/isoc/arch/arm/setjmp_a32.S` where it uses some special function `ftrace_setjmp` and `ftrace_longjmp` when `CFG_FTRACE_SUPPORT` is set ? I tried to hack in `lib/libutils/ext/ftrace/ftrace.c` to get it work but it seams to be either you use everything or nothing here.
Will it be very annoying for me to bundle ftrace ? I just wanna display deltas of time in the bget test of libutils called by xtest, but as long as it doesn't slows everything it should still be usable for me
